src/df/unit_tests/tag_analysis_test.rs

```diff
@@ -0,0 +1,225 @@
+  rules_post_order.reverse();
+  // The entry point of the rule graph
+  let entry_rule = &rules[0];
+  analysis.run_analysis(rules_post_order, vec![entry_rule.clone()]);
+
+  // Check the sigma in of the 2nd rule
+  let sigma = analysis
+    .sigma_in()
+    .get(&rules[2])
+    .unwrap()
+    .variables
+    .clone();
+  let expected = vec!["@method_name", "@class_name", "@detected_class_name"]
+    .into_iter()
+    .map(|s| s.to_string())
+    .collect::<HashSet<String>>();
+  assert_eq!(sigma, expected);
+}
+
+#[test]
+fn test_forward_analysis_simple() {
+  let rules = vec![
+    piranha_rule! {
+      name = "add_inner_class",
+      query = "(
+        (class_declaration name: (_)@class_name
+            body : (class_body ((_)*) @class_members)  @class_body
+        ) @class_declaration
+        (#eq? @class_name \"FooBar\")
+        )"
+    },
+    piranha_rule! {
+    name = "add_field_declaration",
+    query = "(
+        (class_declaration name: (_)@class_name
+            body : (class_body ((_)*) @class_members)  @class_body
+         ) @class_declaration
+        )"
+      },
+  ];
+
+  let edges = vec![edges! {
+    from = "add_inner_class",
+    to = ["add_field_declaration"],
+    scope = "Class"
+  }];
+
+  let graph = RuleGraphBuilder::default()
+    .rules(rules.clone())
+    .edges(edges)
+    .build();
+
+  let forward = ForwardDefiniteAssignment::new(graph, HashSet::new());
+  let mut analysis = DataflowAnalysis::new(forward);
+
+  // Get the rules in post order for optimal performance of the analysis
+  let mut rules_post_order = rules.clone();
+  rules_post_order.reverse();
+  // The entry point of the rule graph
+  let entry_rule = &rules[0];
+  analysis.run_analysis(rules_post_order, vec![entry_rule.clone()]);
+
+  // Check the sigma in of the 2nd rule
+  let sigma = analysis
+    .sigma_in()
+    .get(&rules[1])
+    .unwrap()
+    .variables
+    .clone();
+  let expected = vec![
+    "@class_name",
+    "@class_members",
+    "@class_body",
+    "@class_declaration",
+  ]
+  .into_iter()
+  .map(|s| s.to_string())
+  .collect::<HashSet<String>>();
+  assert_eq!(sigma, expected);
+}
+
+#[test]
+fn test_flow_function() {
+  let rule = piranha_rule! {
+    name = "add_inner_class",
+    query = "(
+        (class_declaration name: (_)@class_name
+            body : (class_body ((_)*) @class_members) @class_body
+        ) @class_declaration
+        (#eq? @class_name_x \"@other_undefined_variable\")
+        (#eq? @class_name_1 \"@_undefined_variable\")
+        )"
+  };
+
+  let edges = vec![];
+
+  let graph = RuleGraphBuilder::default()
+    .rules(vec![rule.clone()])
+    .edges(edges)
+    .build();
+
+  let forward = ForwardDefiniteAssignment::new(graph, HashSet::new());
+
+  let sigma = DefiniteAssignmentSigma {
+    variables: HashSet::new(),
+    is_bottom: false,
+  };
+
+  let new_sigma = forward.flow(&rule, &sigma);
+  assert_eq!(new_sigma.variables.len(), 4);
+}
+
+#[test]
+fn test_flow_function_0() {
```

Comment:
is this a -ve test case? 
can we add more -ve test cases with wrong flow n all ?